tests/integration_python/test_edge_cases.py

```diff
@@ -470,6 +470,362 @@ def test_pypi_url_fragment_in_project_deps(tmp_pixi_workspace: Path, pixi: Path)
+        stderr_contains=["Hash verification failed", "Unsupported hash algorithm"],
+    )
+
+
+def test_pypi_external_source_truncated_hash(tmp_pixi_workspace: Path, pixi: Path) -> None:
+    """Test that truncated hash values fail verification"""
+    pyproject_content = """
+[project]
+version = "0.1.0"
+name = "test"
+requires-python = "== 3.12"
+dependencies = [
+    "requests @ https://files.pythonhosted.org/packages/f9/9b/335f9764261e915ed497fcdeb11df5dfd6f7bf257d4a6a2a686d80da4d54/requests-2.32.3-py3-none-any.whl#sha256=70761cfe03c773ceb22aa2f671b"
+]
+[tool.pixi.workspace]
+platforms = ["linux-64", "osx-arm64", "win-64"]
+channels = ["https://prefix.dev/conda-forge"]
+"""
+    pyproject_path = tmp_pixi_workspace / "pyproject.toml"
+    pyproject_path.write_text(pyproject_content)
+
+    src_dir = tmp_pixi_workspace / "src" / "test"
+    src_dir.mkdir(parents=True, exist_ok=True)
+    (src_dir / "__init__.py").touch()
+
+    # Should fail - truncated hash won't match
+    verify_cli_command(
+        [pixi, "install", "-v"],
+        cwd=tmp_pixi_workspace,
+        expected_exit_code=ExitCode.FAILURE,
+        stderr_contains=["Hash verification failed"],
+    )
+
+
+def test_pypi_external_source_hash_with_params(tmp_pixi_workspace: Path, pixi: Path) -> None:
+    """Test that hash with additional URL parameters works"""
+    pyproject_content = """
+[project]
+version = "0.1.0"
+name = "test"
+requires-python = "== 3.12"
+dependencies = [
+    "requests @ https://files.pythonhosted.org/packages/f9/9b/335f9764261e915ed497fcdeb11df5dfd6f7bf257d4a6a2a686d80da4d54/requests-2.32.3-py3-none-any.whl#sha256=70761cfe03c773ceb22aa2f671b4757976145175cdfca038c02654d061d6dcc6&egg=requests"
+]
+[tool.pixi.workspace]
+platforms = ["linux-64", "osx-arm64", "win-64"]
+channels = ["https://prefix.dev/conda-forge"]
+"""
+    pyproject_path = tmp_pixi_workspace / "pyproject.toml"
+    pyproject_path.write_text(pyproject_content)
+
+    src_dir = tmp_pixi_workspace / "src" / "test"
+    src_dir.mkdir(parents=True, exist_ok=True)
+    (src_dir / "__init__.py").touch()
+
+    # Should succeed - hash is still valid even with additional params
+    verify_cli_command(
+        [pixi, "install", "-v"],
+        cwd=tmp_pixi_workspace,
+        expected_exit_code=ExitCode.SUCCESS,
+    )
+
+
+def test_pypi_external_source_tarball_hash(tmp_pixi_workspace: Path, pixi: Path) -> None:
+    """Test hash verification for source distributions (.tar.gz)"""
+    # Use a real example: charset-normalizer 3.4.0 source distribution
+    pyproject_content = """
+[project]
+version = "0.1.0"
+name = "test"
+requires-python = "== 3.12"
+dependencies = [
+    "charset-normalizer @ https://files.pythonhosted.org/packages/f2/4f/e1808dc01273379acc506d18f1504eb2d299bd4131743b9fc54d7be4df1e/charset_normalizer-3.4.0.tar.gz#sha256=223217c3d4f82c3ac5e29032b3f1c2eb0fb591b72161f86d93f5719079dae93e"
+]
+[tool.pixi.workspace]
+platforms = ["linux-64", "osx-arm64", "win-64"]
```

Comment:
To reduce the overhead of testing we could include only the current platform here.